ingest.py

```diff
@@ -0,0 +1,368 @@
+_DEBUG = _LOGGER.isEnabledFor(logging.DEBUG)
+
+_READINGS_TBL = sa.Table(
+    'readings',
+    sa.MetaData(),
+    sa.Column('asset_code', sa.types.VARCHAR(50)),
+    sa.Column('read_key', sa.types.VARCHAR(50)),
+    sa.Column('user_ts', sa.types.TIMESTAMP),
+    sa.Column('reading', pg.JSONB))
+"""Defines the table that data will be inserted into"""
+
+_CONNECTION_STRING = "dbname='foglamp'"
+
+_STATISTICS_WRITE_FREQUENCY_SECONDS = 5
+
+
+class Ingest(object):
+    """Adds sensor readings to FogLAMP
+
+    Also tracks readings-related statistics.
+    """
+
+    # Class attributes
+    _readings = 0  # type: int
+    """Number of readings accepted before statistics were flushed to storage"""
+
+    _discarded_readings = 0  # type: int
+    """Number of readings rejected before statistics were flushed to storage"""
+
+    _write_statistics_task = None  # type: asyncio.Task
+    """asyncio task for :meth:`_write_statistics`"""
+
+    _write_statistics_sleep_task = None  # type: asyncio.Task
+    """asyncio task for asyncio.sleep"""
+
+    _stop = False
+    """Set to true when the server needs to stop"""
+
+    _started = False
+    """True when the server has been started"""
+
+    _engine = None  # type: aiopg.sa.Engine
+    """Database connection pool"""
+
+    _insert_queue = None  # type: queue.Queue
+    """insert objects are added to this queue"""
+
+    _insert_readings_tasks = None  # type: List[asyncio.Task]
+    """asyncio tasks for :meth:`_insert_readings`"""
+
+    _insert_readings_wait_tasks = None  # type: List[asyncio.Task]
+    """asyncio tasks for asyncio.Queue.get called by :meth:`_insert_readings`"""
+
+    # Configuration
+    _max_db_connections = 1
+    """Maximum number of open database db_connections"""
+
+    _min_inserts_per_transaction = 50
+    """Maximum number of inserts per transaction"""
+
+    _max_inserts_per_transaction = 50
+    """Maximum number of inserts per transaction"""
+
+    _insert_queue_flush_seconds = 1
+    """Number of seconds to wait before flushing the queue"""
+
+    _max_insert_queue_size = 55
+
+    @classmethod
+    def insert_thread(cls):
+        loop = asyncio.new_event_loop()
+        loop.run_forever()
+
+        for _ in range(cls._max_db_connections):
+            cls._insert_readings_wait_tasks.append(None)
+            cls._insert_readings_tasks.append(asyncio.ensure_future(cls._insert_readings(_,
+                                                                    loop=loop)))
+
+        loop.run_until_complete(cls._insert_readings(0, loop))
+
+    @classmethod
+    async def start(cls):
+        """Starts the server"""
+        if cls._started:
+            return
+
+        # Read config
+        """read config
+        _max_db_connections
+        _min_inserts_per_transaction
+        _max_inserts_per_transaction
+         *** validate > 0 ***
+        """
+
+        cls._max_insert_queue_size = 200
```

Comment:
I'd prefer t have this as a constant at the top somewhere, or even something we can override with configuration or in the constructor - but that is icing on the cake.

On moving further down I saw _readings_buffer_size, should this value be that?